File: apps/commerce-streamer/src/main/java/com/loopers/interfaces/consumer/DemoKafkaConsumer.java
package com.loopers.interfaces.consumer;

import com.loopers.config.kafka.KafkaConfig;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.springframework.kafka.annotation.DltHandler;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.kafka.annotation.RetryableTopic;
import org.springframework.kafka.support.Acknowledgment;
import org.springframework.kafka.support.KafkaHeaders;
import org.springframework.messaging.handler.annotation.Header;
import org.springframework.retry.annotation.Backoff;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class DemoKafkaConsumer {

    @KafkaListener(
            topics = "${demo-kafka.test.topic-name}",
            containerFactory = KafkaConfig.BATCH_LISTENER
    )
    @RetryableTopic(
            attempts = "3",
            backoff = @Backoff(delay = 1000L, multiplier = 2.0)
    )
    public void demoListener(
            List<ConsumerRecord<?, ?>> messages,
            Acknowledgment acknowledgment
    ) {
        System.out.println(messages);
        acknowledgment.acknowledge(); // manual ack
    }

    @DltHandler
    public void demoDlt(
            String payload,
            @Header(KafkaHeaders.RECEIVED_TOPIC) String topic,
            @Header(KafkaHeaders.ORIGINAL_TOPIC) String originalTopic,
            @Header(KafkaHeaders.ORIGINAL_PARTITION) Integer originalPartition,
            @Header(KafkaHeaders.ORIGINAL_OFFSET) Long originalOffset
    ) {
        System.out.printf(
                "DLT message received - topic=%s, originalTopic=%s, partition=%s, offset=%s, payload=%s%n",
                topic, originalTopic, String.valueOf(originalPartition), String.valueOf(originalOffset), payload
        );
    }

    @DltHandler
    public void demoDltBatch(
            List<ConsumerRecord<?, ?>> records,
            @Header(KafkaHeaders.RECEIVED_TOPIC) String topic
    ) {
        System.out.printf("DLT batch received - topic=%s, size=%d%n", topic, records != null ? records.size() : 0);
        if (records != null && !records.isEmpty()) {
            ConsumerRecord<?, ?> first = records.get(0);
            System.out.printf(
                    "First record metadata - topic=%s, partition=%d, offset=%d, key=%s, value=%s%n",
                    first.topic(), first.partition(), first.offset(), String.valueOf(first.key()), String.valueOf(first.value())
            );
        }
    }
}

File: apps/commerce-streamer/src/main/kotlin/com/loopers/interfaces/consumer/DemoKafkaConsumer.kt
package com.loopers.interfaces.consumer

import com.loopers.config.kafka.KafkaConfig
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.kafka.support.Acknowledgment
import org.springframework.kafka.annotation.RetryableTopic
import org.springframework.kafka.annotation.DltHandler
import org.springframework.retry.annotation.Backoff
import org.springframework.kafka.support.KafkaHeaders
import org.springframework.messaging.handler.annotation.Header
import org.springframework.stereotype.Component

@Component
class DemoKafkaConsumer {
    @KafkaListener(
        topics = ["\${demo-kafka.test.topic-name}"],
        containerFactory = KafkaConfig.BATCH_LISTENER,
    )
    @RetryableTopic(
        attempts = "3",
        backoff = Backoff(delay = 1000L, multiplier = 2.0)
    )
    fun demoListener(
        messages: List<ConsumerRecord<Any, Any>>,
        acknowledgment: Acknowledgment,
    ) {
        println(messages)
        acknowledgment.acknowledge() // manual ack
    }

    @DltHandler
    fun demoDlt(
        payload: String,
        @Header(KafkaHeaders.RECEIVED_TOPIC) topic: String,
        @Header(KafkaHeaders.ORIGINAL_TOPIC) originalTopic: String?,
        @Header(KafkaHeaders.ORIGINAL_PARTITION) originalPartition: Int?,
        @Header(KafkaHeaders.ORIGINAL_OFFSET) originalOffset: Long?,
    ) {
        println("DLT message received - topic=$topic, originalTopic=$originalTopic, partition=$originalPartition, offset=$originalOffset, payload=$payload")
    }

    @DltHandler
    fun demoDltBatch(
        records: List<ConsumerRecord<Any, Any>>,
        @Header(KafkaHeaders.RECEIVED_TOPIC) topic: String,
    ) {
        println("DLT batch received - topic=$topic, size=${records.size}")
        if (records.isNotEmpty()) {
            val first = records.first()
            println("First record metadata - topic=${first.topic()}, partition=${first.partition()}, offset=${first.offset()}, key=${first.key()}, value=${first.value()}")
        }
    }
}
